### PR TITLE
Initial draft of light client verification process

### PR DESCRIPTION
**Pull request by hcaw**
The following PR is a result of my work on the light client verification process for block inclusion as detailed [here](https://hackmd.io/ohOt4jAPT8uu-soJXHUq0Q).

I've also written up some notes and questions on the implementation which it would be very useful to get some feedback on.

## Questions

- What party is responsible for maintaining the validator set?
- Is there a limit on the total number of validators? This affects the choice
  of variable used for the bitfield
  + If this requirement is known to be static, use some multiple of `uint256`
  + If limit is unbounded, use `bytes`
- What form does `statement` take (am I correct to assume `bytes32`)?
- Could someone more formally specify the verification process required for for the
  initialisation function?
  + Currently unsure as to how the merkle tree structure containing the validators keys
    relates to the signature on `S`
- Can we assume that `msg.sender` == prover?
- What are the minimal parameter requirements for events notifying observers of
  the success of the initial and final validation transactions?
  + Should prefer the more frugal approach of requiring pointers to lookups in
    contract storage as opposed to providing all information within the event
  + A bitfield would suffice for notifying the choice of validators required for the
    finalisation signatures - provided this bitfield can be translated by an observer

## Points for discussion

- Provided the bitfield is used as a key to a lookup table in the validator
  registry contract, there could be a problem in the following situation:
  1. Call to `initialiseValidation()` submitted - sits in mempool
  2. Call to `(un)registerValidator()` submitted - transaction get included in block
     This updates the set of validators into which the bitfield provides a lookup
  3. Transaction at step 1 gets included in block, but now the bitfield which was
     previously provided is invalid
- Abstraction of function input parameters for further development of the
  protocol
  + It may be wise to design the contract interface in such a way as to allow
    for changes in requirements. One such method of abstracting these details
    away is to use a Proxy contract to provide a level of indirection, to which
    one submits a generic `bytes` payload. This payload can be proxied to a contract
    where the data is parsed into discrete variables (provided an agreed upon interface).

### CODE REVIEW

File: contracts/LightClientBridge.sol [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+            validatorRegistry.checkValidatorInSet(),
+            "Error: Sender must be in validator set"
+        );
+
+        require(
+            validateSignature(statement, validatorBitfield, sig, proofs),
+            "Error: Invalid Signature"
+        );
+
+        count = SafeMath.add(count, 1);
```
Comment (denalimarsh) [inline contracts/LightClientBridge.sol:107]: Add the statement `using SafeMath for uint256;` at the top of the contract and then update this to
```suggestion
        count = count.add(1);
```

**Review by denalimarsh — CHANGES_REQUESTED** [review]
Question responses
- The validator set is in Substrate and is maintained on Ethereum by the Bridge. For now, it's okay to assume that the validator set is stored safely on a different contract that you can access via external calls.
- The validator set size is set in Substrate. I'm not sure of the maximum limit off the top of my head, but until we get more info let's assume unbounded and use `bytes`.
- My initial thought is that the right choice for the data packet is `bytes` as it allows us to pass more than 32 bytes of data. The data packet doesn't matter too much right now as we're focused on the validator signature verification scheme at the moment, technically we could even use a uint256 placeholder.
- Will take another pass at this in a bit.
- Can you describe what you mean?
- If I'm understanding correctly, this point is about communicating the contract-selected 1/3rd of validators whose signature we need for the second step. Some potential options:
1. Return the selection directly to the first step's `msg.sender` as a bitfield return value of initialiseValidation(). This is cheap, doesn't require contract storage to save the selection, and gets the data directly to the address that will be sending the 2nd transaction. The problem here is that if the `msg.sender` is negligent and doesn't send the 2nd transaction then that data isn't readily available to other relayers.
2. Store the selection on contract such that other observers could query it. This requires modifying storage, but my primary concern is that other observers wouldn't be aware of an ongoing verification process and that they need to query and use the validator set selection unless there's an event emitted denoting the start of the verification process.
3. Emit an event containing the information. We have to consider how we will manage the verification lifecycle for multiple relayer/observers - an event-driven communication system here would communicate to observers new verification submission, any progress updates/status changes, etc. and will probably massively simplify code on the relayer side. Let's run some gas expenditure tests on the costs of emitting events containing the relevant data, but I'm in favor of this option.

Discussion points
- For the initial implementation let's assume that the validator set remains constant. If there's a change in the validator set mid-verification then we'll need to re-select 1/3rd of validators. This could be done by the ValidatorSet contract pushing the information to this contract and if there are any changes the contract re-selects the new 1/3rd and emits an event containing the (verificationDataHash, newlySelectedValSet) such that the original submitter can still complete the 2nd phase.
- Interesting point, but I think that I'd like to get closer on the initial Solidity implementation before diving too deep into design optimizations at this point.

**Comment by musnit** [comment]
> ## Questions
> * What party is responsible for maintaining the validator set?

 - No one, this shouldn't be owner-based. I think for now basically lets assume that: within the statement payload there will be a way to detect validator set changes. This means that once a new statement is accepted, some logic within a smart contract can dig into that statement and automatically add/remove validators based on the changes within the statement. We could have this logic within the LightClientBridge contract or in the ValidatorRegistry contract or maybe some outer wrapper contract. For now lets put this in the LightClientBridge contract, with some function that happens during acceptBlock() like checkForValidatorSetChanges(statement)

> * Is there a limit on the total number of validators? This affects the choice
>   of variable used for the bitfield

Yeh, though this depends on which network (eg: Kusama and Polkadot are two different networks we may run on). I think Kusama is at about 400 now and Polkadot about 250 at the moment though may change. Long term I think it will still stay under 1000. We can just limit it to that for now and change it later if needed. I've also created this issue for longer term plans re this: https://github.com/Snowfork/polkadot-ethereum/issues/165

> * What form does `statement` take (am I correct to assume `bytes32`)?

Yeh, it'll likely be a keccak256 hash, so bytes32 is safest for now.

> * Could someone more formally specify the verification process required for for the
>   initialisation function?
>   
>   * Currently unsure as to how the merkle tree structure containing the validators keys
>     relates to the signature on `S`

So, I presume that in the ValidatorRegistry we have a merkle root of a merkle tree that contains one leaf for each validator, with that validators public key.

The full logic is then something like:

```solidity
// (This newSignatureCommitment is renamed from initialiseValidation
newSignatureCommitment(
        bytes32 statement,
        uint256 validatorClaimsBitfield,
        bytes32 senderSignatureCommitment,
        MerkleProof senderPublicKeyMerkleProof
    ) 
  {
     //1 - check if senderPublicKeyMerkleProof is valid based on the  ValidatorRegistry merkle root, ie, confirm that the senderPublicKey is from an active validator, with senderPublicKey being returned
    //2 - check if senderSignatureCommitment is correct, ie, check if it matches the signature of senderPublicKey on the statement
    //3 - we're good now, so record statement, validatorClaimsBitfield, senderPublicKey, blocknumber, etc
   _//4 - (only to be done later, lock up sender stake as collateral)_ 
  }
```
> * Can we assume that `msg.sender` == prover?

we could if it simplifies things or is more gas efficient, though we don't have to make this assumption

> * What are the minimal parameter requirements for events notifying observers of
>   the success of the initial and final validation transactions?

lets start with a lot as you have and we can optimize later

>   * Should prefer the more frugal approach of requiring pointers to lookups in
>     contract storage as opposed to providing all information within the event

lets start with it all and we can optimize later. actually we may not even need anything.

>   * A bitfield would suffice for notifying the choice of validators required for the
>     finalisation signatures - provided this bitfield can be translated by an observer
> 

i think we may be able to get away with not even needing anything.

### Here's an idea of the rough pseudocode/process for the entire light client running an entire statement round on both transactions, including offchain logic/ideas:

(before any ethereum transaction)
 - offchain: the sender notices a new statement has been produced with 2/3 majority signatures, including their own. they also notice that there are 250 active validators, and thus that their are at least 167 signatures.
 - offchain: the sender calculates the bitfield based on which validators have signed and stores it
 - offchain: the sender generates their public key merkle proof
 - onchain: the sender submits the 250-bit bitfield (which will have 167 1's and 83 0's, ie, commiting to the 167 specific validators that signed as above), statement, pkmerkleproof and their own signature to newSignatureCommitment. the logic runs as above.
 - offchain: the sender sees the transaction is successful and is included in the chain
 - offchain: the sender assumes that 40 blocks is the preset wait time and that 5 random numbers are needed. these variables can be either hardcoded or could be based on the total number of validators, but is still deterministic and can be calculated offchain without any onchain logic
 - offchain: the sender waits 40+5 = 45 blocks
 - offchain: the sender sees that this 45th block hash blockhash a123h913honasdk
 - offchain: the sender starts a random number generator using the seed a123h913honasdk, and generates 5 random numbers between 1 and 167 that represent the random sample from the set of 250 validators
 -  offchain: the sender extracts the 5 signatures of those 167 from the list of majority signatures
 - onchain: the sender submits the finaliseValidation as below

```solidity
// (This completeSignatureCommitment is renamed from finaliseValidation 
completeSignatureCommitment(
        uint256 id,
        bytes32 statement,
       bytes32[] randomSignatureCommitments,
       uint256[] randomSignatureBitfieldPositions,
       MerkleProof[] randomPublicKeyMerkleProofs

    ) 
  {
     //1 - calculate the random number, with the same offchain logic, ie:
     //      - get statement.blocknumber, add 45, get randomSeedBlockHash, create randomSeed with same randomSeed logic as used offchain, generate 5 random numbers between 1 and 167
    //2 - require random numbers generated onchain match random numbers provided to transaction
    //3 - for each randomSignature, do:
   //      -  take corresponding randomSignatureBitfieldPosition, check with the onchain bitfield that it corresponds to a positive bitfield entry for a validator that did actually sign
   //      -  take corresponding randomSignatureCommitments, check with the onchain bitfield that it corresponds to a positive bitfield entry for a validator that did actually sign
  //     - take corresponding randomPublicKeyMerkleProof, check if it is valid based on the  ValidatorRegistry merkle root, ie, confirm that the randomPublicKey is from an active validator, with randomPublicKey being returned
 //     - take corresponding randomSignatureCommitments,  check if it is correct, ie, check if it matches the signature of randomPublicKey on the statement
   //4 - We're good, we accept the statement
  //5 - we process the statement _(maybe do this in/from another contract, can see later)_
     }
```

> ## Points for discussion
> * Provided the bitfield is used as a key to a lookup table in the validator
>   registry contract, there could be a problem in the following situation:
>   
>   1. Call to `initialiseValidation()` submitted - sits in mempool
>   2. Call to `(un)registerValidator()` submitted - transaction get included in block
>      This updates the set of validators into which the bitfield provides a lookup
>   3. Transaction at step 1 gets included in block, but now the bitfield which was
>      previously provided is invalid

As described above, (un)registerValidator should happen only when the statement is accepted and processed, which should happen atomically in the same transaction as completeSignatureCommitment. There shouldn't be any actual public/external facing  (un)registerValidator function in the end.

> * Abstraction of function input parameters for further development of the
>   protocol
>   
>   * It may be wise to design the contract interface in such a way as to allow
>     for changes in requirements. One such method of abstracting these details
>     away is to use a Proxy contract to provide a level of indirection, to which
>     one submits a generic `bytes` payload. This payload can be proxied to a contract
>     where the data is parsed into discrete variables (provided an agreed upon interface).

ye maybe, lets see later

**Comment by philipstanislaus** [comment]
Added numbers for easier discussion of your questions in the PR description, hope that's fine.

Great answers by @musnit , just a few more thoughts:

7. That should never happen, since changes to validator set are coming through a new commitment/statement. Your function should be internal only.

8. While I like the concept, it also requires some sort of governance mechanism (at least some centralized way to update that target contract of the proxy). We can discuss that longer term, but I'd say it's out of scope for the current iteration of the bridge.

@musnit regarding your process: 

9. I guess 1 seed is enough, and then you just draw n random numbers from that seed.
10. We would need ⅓ of the total validators to submit their signature, in your example 84, not 5. 
11. `randomSignatureBitfieldPositions` is not needed in calldata, the verification will fail when doing `ecrecover`.

**Comment by musnit** [comment]
@philipstanislaus re 10. - how do you get to 1/3? we need to randomly sample n validators such that the probability of sampling a set of colluding dishonest validators is low. anyway i was just using an example for k=5 - this should be parameterized based on statistical analysis and likely change over time.

**Comment by philipstanislaus** [comment]
@musnit re 10. you are correct, misread the spec.

**Comment by musnit** [comment]
@hcaw New additions look good at a glance. Will review in further detail in the coming week. Can you start working on tests for the code in parallel too (even if functionality isn't done/tests don't pass, speccing out the tests should be valuable) - I'll try provide some more suggestions for test cases in the week too.

**Comment by musnit** [comment]
@hcaw Nice changes in the last PR, thanks! Will review in more detail in the next few days. Would be good to get this merged soon cc @denalimarsh @philipstanislaus for reviews.

**Comment by denalimarsh** [comment]
@hcaw I'm getting a TypeScript compilation error on `npx hardhat test` at test/lightClientBridge.ts:130:9. What's the proper way to run the relevant test suite? I can review from a design perspective or try to patch the test suite up, but wanted to check-in and get your recommendation first.

**Comment by musnit** [comment]
@denali@snowfork.com <denali@snowfork.com> npx hardhat compile  fixed that
for me, though getting another second type error now

**Comment by denalimarsh** [comment]
@musnit the Verification/Merkle Tree tests work with the latest changes, but we still need to fix up the Light Client Bridge/Validator Registry tests. They're currently failing, looks like an issue with the proof signer's signature.

**Comment by denalimarsh** [comment]
@musnit the Verification/Merkle Tree tests work with the latest changes, but we still need to fix up the Light Client Bridge/Validator Registry tests. They're currently failing, looks like an issue with the proof signer's signature.

**Comment by musnit** [comment]
yeh, i looked at them too. although 1 of the verification/merkle tree tests
seems pending/not actually working.

**Comment by philipstanislaus** [comment]
Closing this, most changes have been merged through #16 , remainder will be merged through #19

**Pull request closed without merging**